### PR TITLE
[CALCITE-2920] RelBuilder: new method to create an antijoin (Ruben Quesada Lopez)

### PR DESCRIPTION
**Pull request by rubenada**
Add new method to RelBuilder in order to create an antijoin.
Since, for the moment, there is no "AntiJoin" expression, the only possibility to build an AntiJoin is using a LogicalCorrelate with SemiJoinType.ANTI

### CODE REVIEW

**Comment by hsyuan** [comment]
Looks good.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 1]
```diff
@@ -1733,6 +1733,47 @@ public RelBuilder semiJoin(RexNode... conditions) {
     return semiJoin(ImmutableList.copyOf(conditions));
   }
 
+  /** Creates an AntiJoin: since, for the moment, there is no "AntiJoin" expression, the only
+   *  possibility to build an AntiJoin is using a LogicalCorrelate with SemiJoinType.ANTI */
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Can you make this into a comment that is useful to the user? Explain what an ant-join is. A SQL example works well. 

Maybe add a similar description to `semiJoin`.

Description of the limitations of the algebra can be moved to a non-javadoc comment.

File: core/src/test/java/org/apache/calcite/test/RelBuilderTest.java [hunk 2]
```diff
@@ -1428,6 +1428,27 @@ private void project1(int value, SqlTypeName sqlTypeName, String message, String
     assertThat(root, hasTree(expected));
   }
 
+  @Test public void testAntiJoin() {
+    // Equivalent SQL:
+    //   SELECT * FROM dept d WHERE NOT EXISTS (SELECT 1 FROM emp e WHERE e.deptno = d.deptno)
+    final RelBuilder builder = RelBuilder.create(config().build());
+    RelNode root = builder
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/test/RelBuilderTest.java]: Indent should be 4 not 8

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableCorrelateTest.java [hunk 3]
```diff
@@ -133,6 +134,37 @@
...
+                        .antiJoin(
+                                builder.equals(
+                                        builder.field(2, "d", "deptno"),
+                                        builder.field(2, "e", "deptno")))
+                        .project(
+                                builder.field("deptno"),
+                                builder.field("name"))
+                        .sort(builder.field("deptno"))
+                        .build()
+        )
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/test/enumerable/EnumerableCorrelateTest.java]: close parens should not be on their own line

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 1 above)
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: reviewed

**Comment by rubenada** [comment]
Following the comment by @julianhyde in the Jira ticket, I think we could be a bit more ambitious with this PR and build a "true AntiJoin relational expression"; so I'd propose the following:
- Modify `SemiJoin` class: add a new instance field `protected final SemiJoinType semiJoinType;`, which will be by default `SemiJoinType.SEMI` (to ensure backwards compatibility); but that could also be `SemiJoinType.ANTI`, to represent an AntiJoin (a.k.a. AntiSemiJoin)
- We could easily have two implementations of Antijoin:
a) Enumerable: small modification in `EnumerableSemiJoin` and `EnumerableSemiJoinRul`e in order to propagate the `SemiJoinType`, plus a new antiJoin method in `EnumerableDefaults` (basically the same as semiJoin method, but with a "not contains" predicate instead of "contains")
b) Correlate: small modification in `JoinToCorrelateRule` to propagate the `SemiJoin.semiJoinType` will suffice to create the appropriate `LogicalCorrelate` with type SEMI / ANTI
- RelBuilder antiJoin method would no longer create a `LogicalCorrelate` type=ANTI, but instead a `SemiJoin` type=ANTI (using a new method in `SemiJoinFactory` that would accept an additional `SemiJoinType` parameter).
What do you think?

**Comment by julianhyde** [comment]
If the RelBuilder change is done, let's commit it and close the case.

I like the idea of an "anti-join relational expression", but let's discuss it somewhere better than github. Perhaps you should join the "Join, SemiJoin, Correlate" thread on the dev list.

**Comment by rubenada** [comment]
Ok @julianhyde, let's keep this PR as it is, and move the discussion regarding the "anti-join relational expression" to a separate issue.

**Pull request closed without merging**